packages/compass-query-bar/src/components/query-bar/query-bar.tsx

```diff
@@ -49,74 +61,260 @@ const openQueryHistoryStyles = cx(
+    ['collation', 'skip', 'limit', 'maxTimeMS'],
+  ],
+  onApply: _onApply,
+  onChangeQueryOption,
+  onReset: _onReset,
   queryState,
+  refreshEditorAction,
+  schemaFields,
+  serverVersion,
   showQueryHistoryButton = true,
   toggleExpandQueryOptions,
-  toggleQueryHistory,
+  toggleQueryHistory: _toggleQueryHistory,
+  valid: isQueryValid,
+  ...queryOptionProps
 }) => {
-  return (
-    <div className={queryBarStyles}>
-      {showQueryHistoryButton && (
-        <>
-          <Label
-            className={openQueryHistoryLabelStyles}
-            htmlFor="open-query-history"
+  const onReset = useCallback(
+    (evt: React.MouseEvent) => {
+      // Prevent form submission.
+      evt.preventDefault();
+
+      _onReset();
+    },
+    [_onReset]
+  );
+
+  const toggleQueryHistory = useCallback(
+    (evt: React.MouseEvent) => {
+      // Prevent form submission.
+      evt.preventDefault();
+
+      _toggleQueryHistory();
+    },
+    [_toggleQueryHistory]
+  );
+
+  const onApply = useCallback(() => {
+    if (isQueryValid) {
+      _onApply();
+    }
+  }, [_onApply, isQueryValid]);
+
+  const renderQueryOption = useCallback(
+    (queryOption: QueryOption) => {
+      const hasError = !queryOptionProps[`${queryOption}Valid`];
+
+      const placeholder =
+        queryOptionProps[`${queryOption}Placeholder`] ||
+        OPTION_DEFINITION[queryOption].placeholder;
+
+      return (
+        <QueryOptionComponent
+          autoPopulated={autoPopulated}
+          hasError={hasError}
+          key={`query-option-${queryOption}`}
+          queryOption={queryOption}
+          onChange={(value: string) => onChangeQueryOption(queryOption, value)}
+          onApply={onApply}
+          placeholder={placeholder}
+          refreshEditorAction={refreshEditorAction}
+          schemaFields={schemaFields}
+          serverVersion={serverVersion}
+          value={queryOptionProps[`${queryOption}String`]}
+        />
+      );
+    },
+    [
+      autoPopulated,
+      onApply,
+      onChangeQueryOption,
+      refreshEditorAction,
+      schemaFields,
+      serverVersion,
+      queryOptionProps,
+    ]
+  );
+
+  const renderQueryOptionRow = useCallback(
+    (queryOption: string | string[], key: number) => (
+      <div
+        className={cx(rowStyles, key === 0 && firstQueryOptionsRowStyles)}
+        key={key}
+      >
+        {typeof queryOption === 'string'
+          ? renderQueryOption(queryOption as QueryOption)
+          : queryOption.map((optionName: string) =>
+              renderQueryOption(optionName as QueryOption)
+            )}
+        {key === layout.length - 1 && (
+          <Link
+            className={queryDocsLinkStyles}
+            href="https://docs.mongodb.com/compass/current/query/filter/"
+            target="_blank"
           >
-            Query
-          </Label>
+            Learn more
+          </Link>
+        )}
+      </div>
+    ),
+    [renderQueryOption]
+  );
```

Comment:
Maybe would make sense to actually make them into components instead of memoised functions inside render method? This is getting pretty hard to decypher when reading through the code